hopsworks-common/src/main/java/io/hops/hopsworks/common/proxies/CAProxy.java

```diff
@@ -0,0 +1,220 @@
+        }
+      };
+      return retryableAction.tryAction();
+    } catch (JsonProcessingException ex) {
+      throw new HopsSecurityException(RESTCodes.SecurityErrorCode.CSR_ERROR, Level.SEVERE, null, null, ex);
+    } catch (ClientProtocolException ex) {
+      LOG.log(Level.SEVERE, "Could not sign CSR", ex);
+      throw new HopsSecurityException(RESTCodes.SecurityErrorCode.CSR_ERROR, Level.SEVERE, null, null, ex.getCause());
+    } catch (IOException ex) {
+      LOG.log(Level.SEVERE, "Could not sign CSR", ex);
+      throw new GenericException(RESTCodes.GenericErrorCode.UNKNOWN_ERROR, Level.SEVERE,
+          "Generic error while signing CSR", null, ex);
+    }
+  }
+  
+  public void revokeProjectX509(String certificateIdentifier) throws HopsSecurityException, GenericException {
+    revokeX509(certificateIdentifier, CA_PATH.PROJECT_CA_PATH);
+  }
+  
+  public void revokeDelaX509(String certificateIdentifier) throws HopsSecurityException, GenericException {
+    revokeX509(certificateIdentifier, CA_PATH.DELA_CA_PATH);
+  }
+  
+  private void revokeX509(String certificateIdentifier, CA_PATH path) throws HopsSecurityException, GenericException {
+    if (Strings.isNullOrEmpty(certificateIdentifier)) {
+      throw new HopsSecurityException(RESTCodes.SecurityErrorCode.CERTIFICATE_NOT_FOUND, Level.SEVERE,
+          null, "Certificate Identifier cannot be null or empty");
+    }
+    try {
+      URI revokeURI = new URIBuilder(path.path)
+          .addParameter(CERTIFICATE_IDENTIFIER, certificateIdentifier)
+          .build();
+      HttpDelete httpRequest = new HttpDelete(revokeURI);
+      client.setAuthorizationHeader(httpRequest);
+      
+      HttpRetryableAction<Void> retryableAction = new HttpRetryableAction<Void>() {
+        @Override
+        public Void performAction() throws ClientProtocolException, IOException {
+          return client.execute(httpRequest, CA_REVOKE_RESPONSE_HANDLER);
+        }
+      };
+      retryableAction.tryAction();
+    } catch (URISyntaxException ex) {
+      throw new GenericException(RESTCodes.GenericErrorCode.UNKNOWN_ERROR, Level.SEVERE, null, null, ex);
+    } catch (ClientProtocolException ex) {
+      LOG.log(Level.SEVERE, "Could not revoke X.509 " + certificateIdentifier, ex);
+      throw new HopsSecurityException(RESTCodes.SecurityErrorCode.CERTIFICATE_REVOKATION_ERROR, Level.SEVERE,
+          null, null, ex);
+    } catch (IOException ex) {
+      LOG.log(Level.SEVERE, "Could not revoke X.509 " + certificateIdentifier, ex);
+      throw new GenericException(RESTCodes.GenericErrorCode.UNKNOWN_ERROR, Level.SEVERE,
+          "Generic error while revoking X.509", null, ex);
+    }
+  }
+  
+  private class CASignCSRResponseHandler implements ResponseHandler<CSR> {
+  
+    @Override
+    public CSR handleResponse(HttpResponse response) throws ClientProtocolException, IOException {
+      int status = response.getStatusLine().getStatusCode();
```

Comment:
Thanks for taking the time and refactoring the code, looks way better now :) And less duplicated code.
I don't see where you account for an unauthorized (token expired) response. I think in that case we should explicitly trigger the renewal

Reply:
Normally this should never happen. The token will be renewed by `ServiceJWTKeepAlive` If somehow we end-up with an invalid token in the database, then something has stopped working :) If there isn't too much synchronization effort, I'll try to cover this case

Reply:
Yes, but as of right now the `ServiceJWTKeepAlive` runs every 3.5 days or so. Even if you set it to run every hour, you can still get a request for an operation in that hour and the token would be invalid. Or am I missing something?

Reply:
I think you are missing the fact that a token is valid until `expirationDate + leeway`. Yes, if the configuration is wrong (very short leeway period), then with the current implementation you will end-up with an invalid token. But I changed it to run more frequently. The `expirationDate` is mainly for auto-renewable tokens. If `JWTFilter` receives a token and `expirationDate <= CurrentDate < expirationDate+leeway`, then the token is still valid, but it will be renewed automatically.